packages/types/src/type/ExtrinsicEra.ts

```diff
@@ -2,33 +2,224 @@
+  current: number;
+  period: number;
+}
+
+const VALID_IMMORTAL = new U8a([0]);
 
 /**
  * @name ExtrinsicEra
  * @description
  * The era for an extrinsic, indicating either a mortal or immortal extrinsic
  */
-export default class ExtrinsicEra extends U8a {
+export default class ExtrinsicEra extends Enum implements IExtrinsicEra {
+  constructor (value?: any) {
+    super({
+      ImmortalEra,
+      MortalEra
+    }, ExtrinsicEra.decodeExtrinsicEra(value));
+  }
+
+  private static decodeExtrinsicEra (value: MortalMethod | Uint8Array | string = new Uint8Array()): Uint8Array | Object | undefined {
+    if (isHex(value)) {
+      return ExtrinsicEra.decodeExtrinsicEra(hexToU8a(value));
+    } else if (isU8a(value)) {
+      if (!value.length || value[0] === 0) {
+        return new Uint8Array([0, 0]);
+      } else {
+        return new Uint8Array([1, value[0], value[1]]);
+      }
+    } else if (isObject(value)) {
+      return { MortalEra: value };
+    }
+
+    throw new Error('Invalid data passed to Era');
+  }
+
+  /**
+   * @description Overide the encoded length method
+   */
+  get encodedLength (): number {
+    if (this.index === 0) {
+      return this.asImmortalEra.encodedLength;
+    } else {
+      return this.asMortalEra.encodedLength;
+    }
+  }
+
+  /**
+   * @description Returns the item as a [[ImmortalEra]]
+   */
+  get asImmortalEra (): ImmortalEra {
+    assert(this.isImmortalEra, `Cannot convert '${this.type}' via asImmortalEra`);
+
+    return this.value as ImmortalEra;
+  }
+
+  /**
+   * @description Returns the item as a [[MortalEra]]
+   */
+  get asMortalEra (): MortalEra {
+    assert(this.isMortalEra, `Cannot convert '${this.type}' via asMortalEra`);
+
+    return this.value as MortalEra;
+  }
+
+  /**
+   * @description `true` if Immortal
+   */
+  get isImmortalEra (): boolean {
+    return this.index === 0;
+  }
+
+  /**
+   * @description `true` if Mortal
+   */
+  get isMortalEra (): boolean {
+    return this.index > 0;
+  }
+
+  /**
+   * @description Encodes the value as a Uint8Array as per the parity-codec specifications
+   * @param isBare true when the value has none of the type-specific prefixes (internal)
+   */
+  toU8a (isBare?: boolean): Uint8Array {
+    if (this.index === 0) {
+      return this.asImmortalEra.toU8a(isBare);
+    } else {
+      return this.asMortalEra.toU8a(isBare);
+    }
+  }
+}
+
+/**
+ * @name ImmortalEra
+ * @description
+ * The ImmortalEra for an extrinsic
+ */
+export class ImmortalEra extends U8a {
   constructor (value?: AnyU8a) {
-    super(
-      ExtrinsicEra.decodeExtrinsicEra(value)
-    );
+    super(value);
+
+    assert(this.eq(VALID_IMMORTAL), `IMMORTAL: expected ${VALID_IMMORTAL.toHex()}, found ${this.toHex()}`);
+  }
+}
+
+/**
+ * @name MortalEra
+ * @description
+ * The MortalEra for an extrinsic, indicating period and phase
+ */
+export class MortalEra extends Tuple {
+  constructor (value?: any) {
+    super({
+      period: U8,
+      phase: U8
```

Comment:
both of these should extend U64: https://github.com/paritytech/substrate/blob/0e6a407a13fbc5a5ad200645aed72bb8a8e528d7/core/sr-primitives/src/generic/era.rs#L25

Reply:
period and phase are u64 for arithmetic operation and encoded as u8

Reply:
Yes. So decodeMortalEra should take [U8, U8], do the magic decoding, and return [U64, U64] to be passed into the constructor.

Looking at the test now: `expect(extrinsicEra.asMortalEra.phase.toNumber()).toEqual(20000);`, where phase is a U8...

Reply:
> I think we don't have to follow that because in js side they are anyway a BN.

Ah, just read your edit. OK this is debatable, though I still prefer U64 for correctness. If I'm not mistaken, there are only these 2 lines to change, if won't affect anywhere else.

Reply:
my concern is if we make them u64, will we potentially have overflow issues?

Reply:
i think you are right